compiler/rustc_const_eval/src/const_eval/machine.rs

```diff
@@ -457,6 +477,46 @@ impl<'tcx> interpret::Machine<'tcx> for CompileTimeMachine<'tcx> {
                     )?;
                 }
             }
+
+            sym::const_leak => {
+                let ptr = ecx.read_pointer(&args[0])?;
+                let size = ecx.read_scalar(&args[1])?.to_target_usize(ecx)?;
+                let align = ecx.read_scalar(&args[2])?.to_target_usize(ecx)?;
+
+                let size = Size::from_bytes(size);
+                let align = match Align::from_bytes(align) {
+                    Ok(a) => a,
+                    Err(err) => throw_ub_custom!(
+                        fluent::const_eval_invalid_align_details,
+                        name = "const_leak",
+                        err_kind = err.diag_ident(),
+                        align = err.align()
+                    ),
+                };
+
+                // If an allocation is created in an another const,
+                // we don't reallocate it.
+                let (alloc_id, _, _) = ecx.ptr_get_alloc_id(ptr, 0)?;
+                let is_allocated_in_another_const = matches!(
+                    ecx.tcx.try_get_global_alloc(alloc_id),
+                    Some(interpret::GlobalAlloc::Memory(_))
+                );
+
+                if is_allocated_in_another_const {
+                    // just return the pointer that was passed in
+                    ecx.write_pointer(ptr, dest)?;
+                } else {
+                    let ptr = ecx.leak_const_heap_ptr(
+                        ptr,
+                        size,
+                        align,
+                        interpret::MemoryKind::Machine(MemoryKind::Heap),
+                        interpret::MemoryKind::Machine(MemoryKind::HeapLeaked),
```

Comment:
Do we need a separate memory kind? I think the interner will be happy with anything that didn't get deallocated

Reply:
Similar to how it interns stack allocations from the trailing expression